### ISSUE TITLE
GroupBy and Where with != operator couldn't be translated to sql

### ISSUE THREAD

**Issue by meancrazy** [dotnet/efcore#13411]
The LINQ expression 'where ({from StockProductsHistory o in [x] select [o].Sale => Min()} != 100)' could not be translated and will be evaluated locally.

### Steps to reproduce
This code generate sql with group by and having:
```c#
Context.Stocks
       .Where(x => x.Product.Contract.Id == 5783)
       .GroupBy(x => x.Product.IDH)
       .Select(x => new
       {
           x.Key,
           Price = x.Min(o => o.Sale)
       })
       .Where(x => x.Price > 100 || x.Price < 100)
       .ToList()
```
And this code evaluate locally:
```c#
Context.Stocks
       .Where(x => x.Product.Contract.Id == 5783)
       .GroupBy(x => x.Product.IDH)
       .Select(x => new
       {
           x.Key,
           Price = x.Min(o => o.Sale)
       })
       .Where(x => x.Price != 100)
       .ToList()
```

### Further technical details
EF Core version: 2.1.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017 15.8.4

**Comment by ajcvickers** [comment]
@meancrazy Can you post a runnable project/solution or complete code listing that demonstrates the behavior you are seeing?

Note for triage: I attempted to repro this and got an exception in the first case:
```
System.InvalidOperationException: variable 'o' of type 'Microsoft.EntityFrameworkCore.Storage.ValueBuffer' referenced from scope '', but it is not defined
   at System.Linq.Expressions.Compiler.VariableBinder.Reference(ParameterExpression node, VariableStorageKind storage)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitParameter(ParameterExpression node)
   at System.Linq.Expressions.ParameterExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
   at System.Linq.Expressions.BinaryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
   at System.Linq.Expressions.BinaryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda)
   at System.Linq.Expressions.Expression`1.Compile(Boolean preferInterpretation)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateExecutorLambda[TResults]()
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass13_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.Main() in C:\Stuff\TwoOneCore\TwoOneCore\Program.cs:line 89
```
Second case generates:
```sql
SELECT [x].[Id], [x].[ProductId], [x].[Sale], [x.Product].[IDH]
FROM [Stocks] AS [x]
LEFT JOIN [Product] AS [x.Product] ON [x].[ProductId] = [x.Product].[Id]
LEFT JOIN [Contract] AS [x.Product.Contract] ON [x.Product].[ContractId] = [x.Product.Contract].[Id]
WHERE [x.Product.Contract].[Id] = 5783
ORDER BY [x.Product].[IDH]
```

My code:
```C#
public class Stock
{
    public int Id { get; set; }
    public Product Product { get; set; }
    public int Sale { get; set; }
}

public class Product
{
    public int Id { get; set; }
    public Contract Contract { get; set; }
    public int IDH { get; set; }
}

public class Contract
{
    public int Id { get; set; }
}

public class BloggingContext : DbContext
{
    private static readonly LoggerFactory Logger
        = new LoggerFactory(new[] { new ConsoleLoggerProvider((_, __) => true, true) });

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseLoggerFactory(Logger)
            .UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Test;ConnectRetryCount=0");

    public DbSet<Stock> Stocks { get; set; }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
    }
}

public class Program
{
    public static void Main()
    {
        using (var context = new BloggingContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();
        }

        using (var context = new BloggingContext())
        {
            context.Stocks
                .Where(x => x.Product.Contract.Id == 5783)
                .GroupBy(x => x.Product.IDH)
                .Select(x => new
                {
                    x.Key,
                    Price = x.Min(o => o.Sale)
                })
                //.Where(x => x.Price != 100)
                .Where(x => x.Price > 100 || x.Price < 100)
                .ToList();
        }
    }
}
```

**Comment by meancrazy** [comment]
It's so strange, but if you change `Stock` class to this (add only one line):
```csharp
    public class Stock
    {
        public int Id { get; set; }

        public int ProductId { get; set; }

        public Product Product { get; set; }

        public int Sale { get; set; }
    }
```
program will run successfully and generate sql:
```sql
SELECT [x.Product].[IDH] AS [Key], MIN([x].[Sale]) AS [Price]
      FROM [Stocks] AS [x]
      INNER JOIN [Product] AS [x.Product] ON [x].[ProductId] = [x.Product].[Id]
      LEFT JOIN [Contract] AS [x.Product.Contract] ON [x.Product].[ContractId] = [x.Product.Contract].[Id]
      WHERE [x.Product.Contract].[Id] = 5783
      GROUP BY [x.Product].[IDH]
      HAVING (MIN([x].[Sale]) > 100) OR (MIN([x].[Sale]) < 100)
```

So it's a new issue?

**Comment by ajcvickers** [comment]
Note from triage: there are two issues tracked here:
* Client evaluation of the `!=` query when relationship is required (non-nullable FK property)
* Exception from same when relationship is optional (nullable shadow FK property)

**Comment by smitpatel** [comment]
This is fixed in In 3.1
For optional relationship

SQL in query 1
```SQL
      SELECT [p].[IDH] AS [Key], MIN([s].[Sale]) AS [Price]
      FROM [Stocks] AS [s]
      LEFT JOIN [Product] AS [p] ON [s].[ProductId] = [p].[Id]
      LEFT JOIN [Contract] AS [c] ON [p].[ContractId] = [c].[Id]
      WHERE [c].[Id] = 5783
      GROUP BY [p].[IDH]
      HAVING (MIN([s].[Sale]) > 100) OR (MIN([s].[Sale]) < 100)
```

SQL in query 2
```SQL
      SELECT [p].[IDH] AS [Key], MIN([s].[Sale]) AS [Price]
      FROM [Stocks] AS [s]
      LEFT JOIN [Product] AS [p] ON [s].[ProductId] = [p].[Id]
      LEFT JOIN [Contract] AS [c] ON [p].[ContractId] = [c].[Id]
      WHERE [c].[Id] = 5783
      GROUP BY [p].[IDH]
      HAVING (MIN([s].[Sale]) <> 100) OR MIN([s].[Sale]) IS NULL
```

For required relationship
SQL in query 1
```SQL
      SELECT [p].[IDH] AS [Key], MIN([s].[Sale]) AS [Price]
      FROM [Stocks] AS [s]
      INNER JOIN [Product] AS [p] ON [s].[ProductId] = [p].[Id]
      LEFT JOIN [Contract] AS [c] ON [p].[ContractId] = [c].[Id]
      WHERE [c].[Id] = 5783
      GROUP BY [p].[IDH]
      HAVING (MIN([s].[Sale]) > 100) OR (MIN([s].[Sale]) < 100)
```

SQL in query 2
```SQL
      SELECT [p].[IDH] AS [Key], MIN([s].[Sale]) AS [Price]
      FROM [Stocks] AS [s]
      INNER JOIN [Product] AS [p] ON [s].[ProductId] = [p].[Id]
      LEFT JOIN [Contract] AS [c] ON [p].[ContractId] = [c].[Id]
      WHERE [c].[Id] = 5783
      GROUP BY [p].[IDH]
      HAVING (MIN([s].[Sale]) <> 100) OR MIN([s].[Sale]) IS NULL
```

Assigning to @maumar to add regression tests.
